README.md
# 说明

## 注意

> 工具依赖 `changelogen` 的命令行，如果使用发版功能之前未安装，则会自动全局安装最新版本。如有意外情况，请手动安装，**changelogen 当前为 0.5.5**

## 安装

```bash
npm i young-commit -g
```

## 基础使用

- 将修改添加到暂存区之后
- 命令行输入 `young-commit` 或者别名 `yc`

## 带参的快捷操作

```bash
USAGE yc [OPTIONS] 
# USAGE young-commit [OPTIONS] 

OPTIONS

     -i, --init    首次提交    
  -r, --release    版本发布    
  -p, --publish    发布到 npm 
  -v, --version    版本信息    
     -h, --help    展示工具使用方法
```

## 灵感来源

[bumpp](https://github.com/antfu/bumpp)

[changelogen](https://github.com/unjs/changelogen)

CHANGELOG.md
# Changelog


## v3.0.2

[compare changes](https://github.com/BluesYoung-web/young-commit/compare/v3.0.1...v3.0.2)

### 🔥 Performance

- 🔥 优化安装体验 ([2d68c40](https://github.com/BluesYoung-web/young-commit/commit/2d68c40))

### 🩹 Fixes

- ⚠️  🩹 fix typo ([0132195](https://github.com/BluesYoung-web/young-commit/commit/0132195))

#### ⚠️ Breaking Changes

- ⚠️  🩹 fix typo ([0132195](https://github.com/BluesYoung-web/young-commit/commit/0132195))

### ❤️ Contributors

- 张扬 <bluesyoung_web@163.com>

## v3.0.1

[compare changes](https://github.com/BluesYoung-web/young-commit/compare/v3.0.0...v3.0.1)

### 🩹 Fixes

- 🩹 bug fix ([091b0ff](https://github.com/BluesYoung-web/young-commit/commit/091b0ff))

### ❤️ Contributors

- 张扬 <bluesyoung_web@163.com>

## v3.0.0

[compare changes](https://github.com/BluesYoung-web/young-commit/compare/v2.4.3...v3.0.0)

### 🚀 Enhancements

- 🚀 新增发布至 npm 的功能 ([8b48807](https://github.com/BluesYoung-web/young-commit/commit/8b48807))

### 🩹 Fixes

- 🩹 bug fix ([5c736fd](https://github.com/BluesYoung-web/young-commit/commit/5c736fd))

### 💅 Refactors

- ⚠️  💅 代码结构拆分 + 功能新增 ([5b884b1](https://github.com/BluesYoung-web/young-commit/commit/5b884b1))

### 📖 Documentation

- 📖 update readme ([4d398a7](https://github.com/BluesYoung-web/young-commit/commit/4d398a7))

#### ⚠️ Breaking Changes

- ⚠️  💅 代码结构拆分 + 功能新增 ([5b884b1](https://github.com/BluesYoung-web/young-commit/commit/5b884b1))

### ❤️ Contributors

- 张扬 <bluesyoung_web@163.com>

## v2.4.3

[compare changes](https://github.com/BluesYoung-web/young-commit/compare/v2.4.2...v2.4.3)

### 🩹 Fixes

- 🩹 文件加入白名单 ([9f2b274](https://github.com/BluesYoung-web/young-commit/commit/9f2b274))

### ❤️ Contributors

- 张扬 <bluesyoung_web@163.com>

## v2.4.2

[compare changes](https://github.com/BluesYoung-web/young-commit/compare/v2.4.1...v2.4.2)

### 🩹 Fixes

- 🩹 fix windows exec bug ([7661460](https://github.com/BluesYoung-web/young-commit/commit/7661460))

### ❤️ Contributors

- 张扬 <bluesyoung_web@163.com>

## v2.4.1

[compare changes](https://github.com/BluesYoung-web/young-commit/compare/v2.4.0...v2.4.1)

### 🩹 Fixes

- 🩹 try to fix bug when install to use ([3fa30e2](https://github.com/BluesYoung-web/young-commit/commit/3fa30e2))

### ❤️ Contributors

- 张扬 <bluesyoung_web@163.com>

## v2.4.0

[compare changes](https://github.com/BluesYoung-web/young-commit/compare/v2.3.1...v2.4.0)

### 🚀 Enhancements

- 🚀 兼容 window 环境(which polyfill) ([26dca13](https://github.com/BluesYoung-web/young-commit/commit/26dca13))

### 🏡 Chore

- ⚠️  🏡 change eslint as code format tool & code format ([2fcc99c](https://github.com/BluesYoung-web/young-commit/commit/2fcc99c))

#### ⚠️ Breaking Changes

- ⚠️  🏡 change eslint as code format tool & code format ([2fcc99c](https://github.com/BluesYoung-web/young-commit/commit/2fcc99c))

### ❤️ Contributors

- 张扬 <bluesyoung_web@163.com>

## v2.3.1

[compare changes](https://github.com/BluesYoung-web/young-commit/compare/v2.3.0...v2.3.1)

### 🩹 Fixes

- 🩹 Ctrl + C 需要终止操作 ([2fb9383](https://github.com/BluesYoung-web/young-commit/commit/2fb9383))

### ❤️ Contributors

- 张扬 <bluesyoung_web@163.com>

## v2.3.0

[compare changes](https://github.com/BluesYoung-web/young-commit/compare/v2.2.0...v2.3.0)

### 🩹 Fixes

- 🩹 兼容未安装过 changelogen 的情况 ([1a0f1b0](https://github.com/BluesYoung-web/young-commit/commit/1a0f1b0))

### 📖 Documentation

- 📖 update README ([27fa02c](https://github.com/BluesYoung-web/young-commit/commit/27fa02c))

### ❤️ Contributors

- 张扬 <bluesyoung_web@163.com>

## v2.2.0

[compare changes](https://github.com/BluesYoung-web/young-commit/compare/v2.1.1...v2.2.0)

### 🚀 Enhancements

- 🚀 兼容 a.b.c.alpha.n 的情况 ([bb80085](https://github.com/BluesYoung-web/young-commit/commit/bb80085))

### ❤️ Contributors

- 张扬 <bluesyoung_web@163.com>

## v2.1.1

[compare changes](https://github.com/BluesYoung-web/young-commit/compare/v2.1.0...v2.1.1)

### 🩹 Fixes

- 🩹 fix init commit bug ([c1d0e0e](https://github.com/BluesYoung-web/young-commit/commit/c1d0e0e))

### ❤️ Contributors

- 张扬 <bluesyoung_web@163.com>

## v2.1.0

[compare changes](https://github.com/BluesYoung-web/young-commit/compare/v2.0.2...v2.1.0)

### 🚀 Enhancements

- 🚀 add other cmd ([3bacd61](https://github.com/BluesYoung-web/young-commit/commit/3bacd61))

### ❤️ Contributors

- 张扬 <bluesyoung_web@163.com>

## v2.0.2

[compare changes](https://github.com/BluesYoung-web/young-commit/compare/v2.0.1...v2.0.2)

### 🩹 Fixes

- 🐛  try to fix use bug ([fcf0599](https://github.com/BluesYoung-web/young-commit/commit/fcf0599))

### ❤️ Contributors

- 张扬 <bluesyoung_web@163.com>

## v2.0.1

[compare changes](https://github.com/BluesYoung-web/young-commit/compare/v2.0.0...v2.0.1)

### 🩹 Fixes

- 🐛  try to fix can't use bug ([1253d4e](https://github.com/BluesYoung-web/young-commit/commit/1253d4e))

### ❤️ Contributors

- 张扬 <bluesyoung_web@163.com>

## v2.0.0

[compare changes](https://github.com/BluesYoung-web/young-commit/compare/v1.0.2...v2.0.0)

### 🚀 Enhancements

- 🚀 新增版本发布功能 ([79fba5b](https://github.com/BluesYoung-web/young-commit/commit/79fba5b))

### 💅 Refactors

- ⚠️  💅 代码重构 ([f35314c](https://github.com/BluesYoung-web/young-commit/commit/f35314c))

### 📖 Documentation

- 📖 add license ([d7c46c7](https://github.com/BluesYoung-web/young-commit/commit/d7c46c7))
- 📖 update readme ([b532b44](https://github.com/BluesYoung-web/young-commit/commit/b532b44))

### 🏡 Chore

- 🏡 rom config for code format ([c70f011](https://github.com/BluesYoung-web/young-commit/commit/c70f011))

#### ⚠️ Breaking Changes

- ⚠️  💅 代码重构 ([f35314c](https://github.com/BluesYoung-web/young-commit/commit/f35314c))

### ❤️ Contributors

- 张扬 <bluesyoung_web@163.com>



cli.js
#! /usr/bin/env node
const { program } = require('commander');
const json = require('./package.json');
const inquirer = require('inquirer');

const shell = require('shelljs');
 
if (!shell.which('git')) {
  shell.echo('Sorry, this script requires git');
  shell.exit(1);
}
 
program.version(json.version)
  .option('-i, --init')
  .parse(process.argv);
const { init } = program.opts();

if (init) {
  const cmd = `git commit -m"init: :tada: 项目初始化"`;
  shell.exec(cmd);
  return;
}

const optionList = [
  {
    name: 'type',
    type: 'list',
    message: '请选择你要提交的类型：',
    choices: [
      'feat     -> 新特性/功能',
      'fix      -> Bug 修复',
      'refactor -> 代码重构',
      'docs     -> 文档修改',
      'style    -> 代码格式修改，非 css',
      'test     -> 测试用例修改',
      'chore    -> 构建或者依赖的修改'
    ],
    filter: (val) => {
      return val.match(/\w+/)[0];
    }
  },
  {
    name: 'use_default_icon',
    type: 'confirm',
    message: '是否使用默认的 gitmoji ?'
  },
  {
    name: 'icon',
    type: 'input',
    message: '请输入你想使用的 gitmoji（只需名称，无需引号）：',
    when: ({ use_default_icon }) => {
      return !use_default_icon;
    },
    validate: (val) => {
      if (!val.indexOf(':') > -1) {
        return true;
      }
      return '只需名称，无需引号';
    }
  },
  {
    name: 'msg',
    type: 'input',
    message: '请输入此次提交的描述：',
    validate: (val) => {
      if (val.trim() !== '') {
        return true;
      }
      return '请勿输入空字符串！！！';
    }
  }
];

inquirer.prompt(optionList).then((answer) => {
  const { type = 'feat', icon, msg } = answer;
  const defaultIcons = {
    feat: 'sparkles',
    fix: 'bug',
    refactor: 'recycle',
    docs: 'memo',
    style: 'lipstick',
    test: 'white_check_mark',
    chore: 'hammer'
  }
  let gitmoji = '';
  if (icon) {
    gitmoji = icon;
  } else {
    gitmoji = defaultIcons[type];
  }
  const command = `git commit -m"${type}: :${gitmoji}: ${msg}"`;
  shell.exec(command);
}).catch(console.error);


src/index.ts
/*
 * @Author: zhangyang
 * @Date: 2023-09-05 08:30:03
 * @LastEditTime: 2023-09-07 10:45:37
 * @Description:
 */
import { readFile } from 'node:fs/promises'
import { defineCommand, runMain, showUsage } from 'citty'
import { $ } from 'execa'
import which from 'which'
import { commit, init, publish, release } from './core'

const main = defineCommand({
  meta: new Promise(async (resolve) => {
    // 编译时，相对路径下的 package.json
    const info = await readFile(new URL('../package.json', import.meta.url), {
      encoding: 'utf-8',
    })
    resolve(JSON.parse(info))
  }),
  args: {
    init: {
      type: 'boolean',
      alias: 'i',
      description: '首次提交',
    },
    release: {
      type: 'boolean',
      alias: 'r',
      description: '版本发布',
    },
    publish: {
      type: 'boolean',
      alias: 'p',
      description: '发布到 npm',
    },
    version: {
      type: 'boolean',
      alias: 'v',
      description: '版本信息',
    },
    help: {
      type: 'boolean',
      alias: 'h',
      description: '展示工具使用方法',
    },
  },
  async run({ args }) {
    try {
      await which('git')
    }
    catch (error) {
      console.error('找不到 git 的安装位置')
      process.exit(1)
    }

    await $`git config --global core.unicode true`

    if (args.version) {
      // @ts-expect-error
      const verison = (await main.meta).version
      console.log(`v${verison}`)
      process.exit(0)
    }

    if (args.help) {
      await showUsage(main)
      process.exit(0)
    }

    if (args.init) {
      await init()
      process.exit(0)
    }

    if (args.release) {
      await release()
      process.exit(0)
    }

    if (args.publish) {
      await publish()
      process.exit(0)
    }

    await commit()
  },

})

runMain(main)


src/core/release.ts
/*
 * @Author: zhangyang
 * @Date: 2023-09-05 10:48:22
 * @LastEditTime: 2023-09-07 13:18:55
 * @Description:
 */
import { readFile, writeFile } from 'node:fs/promises'
import { resolve } from 'node:path'
import { getLastGitTag } from 'changelogen'
import c from 'picocolors'
import prompts from 'prompts'
import semver, { clean as cleanVersion, valid as isValidVersion } from 'semver'
import type { ReleaseType } from 'semver'
import { $, execa } from 'execa'
import task from 'tasuku'
import which from 'which'

/**
 * Returns the next version number for all release types.
 */
function getNextVersions(
  oldVersion: string,
  preid = 'alpha',
): Record<ReleaseType | 'next', string> {
  const next: Record<string, string> = {}

  // 兼容 a.b.c.alpha.n 的情况，标准版本为 a.b.c-alpha.n
  const IsPointPreid = /\.([^0-9\.]+)/gim.test(oldVersion)

  if (IsPointPreid)
    oldVersion = oldVersion.replace(/\.([^0-9\.]+)/gim, '-$1')

  const parse = semver.parse(oldVersion)
  if (typeof parse?.prerelease[0] === 'string')
    preid = parse?.prerelease[0] || 'preid'

  for (const type of [
    'premajor',
    'preminor',
    'prepatch',
    'prerelease',
    'major',
    'minor',
    'patch',
  ] as ReleaseType[])
    next[type] = semver.inc(oldVersion, type, preid)!

  next.next = parse?.prerelease?.length
    ? semver.inc(oldVersion, 'prerelease', preid)!
    : semver.inc(oldVersion, 'patch')!

  if (IsPointPreid) {
    for (const key in next)
      next[key] = next[key].replace(/-([^0-9\.]+)/gim, '.$1')
  }

  return next
}

export async function release() {
  const tag = ((await getLastGitTag()) || 'v0.0.0').toLocaleLowerCase()
  if (tag.indexOf('v') !== 0)
    throw new Error('上一个 tag 不合法')

  const oldVersion = tag.substring(1)

  const next = getNextVersions(oldVersion)
  const PADDING = 13
  const answers = (await prompts([
    {
      type: 'autocomplete',
      name: 'release',
      message: `当前版本 ${c.green(oldVersion)}`,
      initial: 'next',
      choices: [
        { value: 'major', title: `${'major'.padStart(PADDING, ' ')} ${c.bold(next.major)}` },
        { value: 'minor', title: `${'minor'.padStart(PADDING, ' ')} ${c.bold(next.minor)}` },
        { value: 'patch', title: `${'patch'.padStart(PADDING, ' ')} ${c.bold(next.patch)}` },
        { value: 'next', title: `${'next'.padStart(PADDING, ' ')} ${c.bold(next.next)}` },
        {
          value: 'prepatch',
          title: `${'pre-patch'.padStart(PADDING, ' ')} ${c.bold(next.prepatch)}`,
        },
        {
          value: 'preminor',
          title: `${'pre-minor'.padStart(PADDING, ' ')} ${c.bold(next.preminor)}`,
        },
        {
          value: 'premajor',
          title: `${'pre-major'.padStart(PADDING, ' ')} ${c.bold(next.premajor)}`,
        },
        { value: 'none', title: `${'as-is'.padStart(PADDING, ' ')} ${c.bold(oldVersion)}` },
        { value: 'custom', title: 'custom ...'.padStart(PADDING + 4, ' ') },
      ],
    },
    {
      type: prev => (prev === 'custom' ? 'text' : null),
      name: 'custom',
      message: '请输入版本号，例如: a.b.c',
      initial: oldVersion,
      validate: (custom: string) => {
        return isValidVersion(custom) ? true : 'That\'s not a valid version number'
      },
    },
    {
      type: 'confirm',
      name: 'changePackageVersion',
      message: '是否修改 package.json',
      initial: false,
    },
  ], {
    onCancel: () => {
      console.log('------------------主动终止------------------')
      process.exit(0)
    },
  })) as {
    release: ReleaseType | 'next' | 'none' | 'custom'
    custom?: string
    changePackageVersion?: boolean
  }

  const newVersion
    = answers.release === 'none'
      ? oldVersion
      : answers.release === 'custom'
        ? cleanVersion(answers.custom!)!
        : next[answers.release]

  if (!newVersion)
    process.exit(1)

  await task.group(task => [
    task('准备 changelogen 的环境', async () => {
      try {
        await which('changelogen')
      }
      catch (error) {
        await $`npm i -g changelogen`
      }
    }),

    task('生成 CHANGELOG.md', async () => {
      await $`changelogen -r ${newVersion} --output`
    }),
    task('git commit & git tag', async ({ task }) => {
      if (answers.changePackageVersion) {
        // 运行时，当前运行目录下的 package.json
        const pkgPath = resolve(process.cwd(), './package.json')

        const file = await readFile(pkgPath, { encoding: 'utf-8' })
        const json = JSON.parse(file)
        json.version = newVersion
        await writeFile(pkgPath, JSON.stringify(json, null, 2), { encoding: 'utf-8' })
      }

      await task.group(t => [
        t('git add .', async () => {
          await $`git add .`
        }),
        t('git commit', async () => {
          await execa('git', ['commit', '-m', `chore(release): v${newVersion}`])
        }),
        t('git tag', async () => {
          await execa('git', ['tag', '-am', `chore: 🏡 release v${newVersion}`, `v${newVersion}`])
        }),
      ])
    }),
  ])
}


src/core/commit.ts
/*
 * @Author: zhangyang
 * @Date: 2023-09-07 10:42:42
 * @LastEditTime: 2023-09-21 11:25:16
 * @Description:
 */
import { execa } from 'execa'
import prompts from 'prompts'

export async function commit() {
  const CommitTypeMap = new Map<string, string>()
  CommitTypeMap.set('feat', '🚀 Enhancements')
  CommitTypeMap.set('perf', '🔥 Performance')
  CommitTypeMap.set('fix', '🩹 Fixes')
  CommitTypeMap.set('refactor', '💅 Refactors')
  CommitTypeMap.set('docs', '📖 Documentation')
  CommitTypeMap.set('build', '📦 Build')
  CommitTypeMap.set('types', '🌊 Types')
  CommitTypeMap.set('chore', '🏡 Chore')
  CommitTypeMap.set('examples', '🏀 Examples')
  CommitTypeMap.set('style', '🎨 Styles')
  CommitTypeMap.set('test', '✅ Tests')
  CommitTypeMap.set('ci', '🤖 CI')

  const answers = (await prompts([
    {
      type: 'autocomplete',
      name: 'type',
      message: '请选择你要提交的类型：',
      initial: 'feat',
      choices: Array.from(CommitTypeMap.entries()).map(([value, title]) => ({
        title: `${value}: ${title}`,
        value,
      })),
    },
    {
      type: 'confirm',
      name: 'breaking',
      initial: false,
      message: '是否为破坏性变更',
    },
    {
      type: 'text',
      name: 'msg',
      message: '请输入此次提交的描述：',
      validate: (val) => {
        if (val.trim() !== '')
          return true

        return '请勿输入空字符串！！！'
      },
    },
  ], {
    onCancel: () => {
      console.log('------------------主动终止------------------')
      process.exit(0)
    },
  })) as {
    type: string
    breaking: boolean
    msg: string
  }

  const { type, breaking, msg } = answers
  const icon = CommitTypeMap.get(type).match(/.+\s/)[0]

  await execa('git', ['commit', '-m', `${type}${breaking ? '!' : ''}: ${icon}${msg}`])
}
